{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e79cd437",
   "metadata": {},
   "outputs": [],
   "source": [
    "import allel\n",
    "from allel.stats import fst, diversity\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess\n",
    "import itertools\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "941d3d5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_variants_for_fst(genotype, subpops: dict) -> allel.GenotypeArray:\n",
    "    pop1, pop2 = list(subpops.keys())\n",
    "    acs = genotype.count_alleles_subpops(subpops)\n",
    "    acu = allel.AlleleCountsArray(acs[pop1][:] + acs[pop2][:])\n",
    "    \n",
    "    flt = acu.is_segregating() & (acu.max_allele() == 1)\n",
    "    print('retaining', np.count_nonzero(flt), 'SNPs')\n",
    "    \n",
    "    filtered_genotype = genotype.compress(flt, axis=0)\n",
    "\n",
    "    return filtered_genotype\n",
    "\n",
    "def dxy(genotype, subpops):\n",
    "    pop1, pop2 = list(subpops.keys())\n",
    "    acs = genotype.count_alleles_subpops(subpops)\n",
    "    pairwise_diff = diversity.mean_pairwise_difference_between(acs[pop1], acs[pop2])\n",
    "\n",
    "    return pairwise_diff\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a77838a9",
   "metadata": {},
   "source": [
    "Extract samples from Mexican populations, excluding cross-contaminated samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "909f9378",
   "metadata": {},
   "outputs": [],
   "source": [
    "pops_file = \"../data/samples/populations.txt\"\n",
    "admixed = \"../data/var/admixture/admixed_individuals.csv\"\n",
    "mex = [\"Amanitaspjack3\", \"Amanitaspjack5\", \"Amanitaspjack6\"]\n",
    "pops = pd.read_table(pops_file, header=None, names=[\"sample\", \"population\"])\n",
    "admixed = pd.read_csv(admixed)\n",
    "pops_mex = pops[pops[\"population\"].isin(mex) & ~pops[\"sample\"].isin(admixed[\"sample\"])]\n",
    "pops_mex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efcf120d",
   "metadata": {},
   "outputs": [],
   "source": [
    "csv_samples = \",\".join(pops_mex[\"sample\"].tolist())\n",
    "csv_samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3355ce1a",
   "metadata": {},
   "source": [
    "Create a VCF file exclusive to Mexican samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a7d90d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_file = \"../data/var/merged_variants.vcf.gz\"\n",
    "output_file = \"../data/var/mex_variants.vcf.gz\"\n",
    "\n",
    "cmd_subset_samples = [\"bcftools\", \"view\", \"-s\", f\"{csv_samples}\", vcf_file]\n",
    "cmd_remove_info = [\"bcftools\", \"annotate\", \"--remove\", \"INFO\"]\n",
    "cmd_refill_tags = [\"bcftools\", \"+fill-tags\", \"--\", \"-t\", \"all\"]\n",
    "cmd_keep_full_callset = [\n",
    "    \"bcftools\", \"view\", \n",
    "    \"-i\", \"INFO/F_MISSING = 0 & AF > 0 & AF < 1\",\n",
    "    \"-m\", \"2\", \n",
    "    \"-M\", \"2\", \n",
    "    \"-v\", \"snps\", \n",
    "    \"-Oz\", \n",
    "    \"-o\", output_file,\n",
    "    \"--write-index\",\n",
    "]\n",
    "\n",
    "process1 = subprocess.Popen(cmd_subset_samples, stdout=subprocess.PIPE)\n",
    "process2 = subprocess.Popen(cmd_remove_info, stdin=process1.stdout, stdout=subprocess.PIPE)\n",
    "process1.stdout.close() # Allow process1 to receive a SIGPIPE if process2 exits\n",
    "process3 = subprocess.Popen(cmd_refill_tags, stdin=process2.stdout, stdout=subprocess.PIPE)\n",
    "process2.stdout.close()\n",
    "process4 = subprocess.Popen(cmd_keep_full_callset, stdin=process3.stdout)\n",
    "process3.stdout.close()\n",
    "# wait for the last process to finish\n",
    "process4.communicate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d77376e",
   "metadata": {},
   "outputs": [],
   "source": [
    "callset = allel.read_vcf(output_file)\n",
    "samples = callset['samples']\n",
    "gt = allel.GenotypeArray(callset['calldata/GT'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b5afe71",
   "metadata": {},
   "outputs": [],
   "source": [
    "pop_indices = {\n",
    "    pop: np.where(np.isin(samples, pops_mex[pops_mex[\"population\"] == pop][\"sample\"]))[0]\n",
    "    for pop in pops_mex[\"population\"].unique()\n",
    "}\n",
    "pop_indices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20bf4f6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "pop_combo = list(itertools.combinations(pops_mex[\"population\"].unique(), 2))\n",
    "pop_combo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19df65ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "fsts_df = pd.DataFrame()\n",
    "for pop1, pop2 in pop_combo:\n",
    "    subpops_idx = {pop: pop_indices[pop] for pop in (pop1, pop2)}\n",
    "    flt_gt = filter_variants_for_fst(gt, subpops_idx)\n",
    "    acs = flt_gt.count_alleles_subpops(subpops_idx)\n",
    "\n",
    "    a, b, c = fst.weir_cockerham_fst(\n",
    "        flt_gt, \n",
    "        subpops=[subpops_idx[pop1], \n",
    "                 subpops_idx[pop2]], \n",
    "        max_allele=1,\n",
    "        blen=1000,\n",
    "    )\n",
    "    \n",
    "    with np.errstate(divide='ignore', invalid='ignore'):\n",
    "        fst_per_variant = (a / (a + b + c))[:, 0]\n",
    "    fst_value = np.nanmean([val for val in fst_per_variant if not np.isnan(val)])\n",
    "    fst_global = (sum(a) / sum(a + b + c))[0]\n",
    "    \n",
    "    num, den = fst.hudson_fst(acs[pop1], acs[pop2])\n",
    "    snp_fst_hudson = num / den\n",
    "    global_fst_hudson = sum(num) / sum(den)\n",
    "    fst_value_hudson = np.nanmean([val for val in snp_fst_hudson if not np.isnan(val)])\n",
    "    \n",
    "    print(f\"Mean Weir-Cockerham's Fst between {pop1} and {pop2}: {fst_value:.4f}\")\n",
    "    print(f\"Global Weir-Cockerham's Fst between {pop1} and {pop2}: {fst_global:.4f}\")\n",
    "    print(f\"Mean Hudson's Fst between {pop1} and {pop2}: {fst_value_hudson:.4f}\")\n",
    "    print(f\"Global Hudson's Fst between {pop1} and {pop2}: {global_fst_hudson:.4f}\")\n",
    "    df = pd.DataFrame({\n",
    "        \"fst_per_variant_wc\": fst_per_variant,\n",
    "        \"fst_per_variant_hudson\": snp_fst_hudson,\n",
    "    })\n",
    "    df[\"fst_global_wc\"] = fst_global\n",
    "    df[\"fst_global_hudson\"] = global_fst_hudson\n",
    "    df[\"populations\"] = f\"{pop1} vs {pop2}\"\n",
    "\n",
    "    fsts_df = pd.concat([fsts_df, df], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec401168",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.boxplot(data=fsts_df, x=\"populations\", y=\"fst_per_variant_wc\")\n",
    "plt.ylabel(\"Fst per variant (Weir-Cockerham)\")\n",
    "plt.xlabel(\"Population Pair\")\n",
    "plt.title(\"Per-variant Weir-Cockerham Fst distribution\")\n",
    "plt.xticks(rotation=30)\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "413d5539",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.boxplot(data=fsts_df, x=\"populations\", y=\"fst_per_variant_hudson\")\n",
    "plt.ylabel(\"Fst per variant (Hudson)\")\n",
    "plt.xlabel(\"Population Pair\")\n",
    "plt.title(\"Per-variant Hudson's Fst distribution\")\n",
    "plt.xticks(rotation=30)\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eff57bb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "dxy_df = pd.DataFrame()\n",
    "for pop_pair in pop_combo:\n",
    "    subpops_idx = {pop: pop_indices[pop] for pop in pop_pair}\n",
    "    per_site_diff = dxy(gt, subpops_idx)\n",
    "    print(f\"{pop_pair}: {np.mean(per_site_diff)}\")\n",
    "    dxy_df = pd.concat((dxy_df, pd.DataFrame({\n",
    "        \"dxy\": per_site_diff,\n",
    "        \"populations\": f\"{pop_pair[0]} vs {pop_pair[1]}\",\n",
    "    })), ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6eef2ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.boxplot(data=dxy_df, x=\"populations\", y=\"dxy\")\n",
    "plt.ylabel(\"Dxy per variant\")\n",
    "plt.xlabel(\"Population Pair\")\n",
    "plt.title(\"Per-variant Dxy distribution\")\n",
    "plt.xticks(rotation=30)\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "conda",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
