{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9661303d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "from jacksonii_analyses import vcf_parser, clustering\n",
    "\n",
    "vcf_path = \"../data/var/filtered_variants.vcf.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbc40e7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from importlib import reload\n",
    "reload(vcf_parser)\n",
    "reload(clustering)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14f1ad74",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = vcf_parser.vcf_to_chr_pos_df(vcf_path)\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf692720",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cluster variants by distance with a threshold of 500 bp\n",
    "dcluster_data = clustering.cluster_variants_by_distance(data, 500)\n",
    "dcluster_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e39091df",
   "metadata": {},
   "outputs": [],
   "source": [
    "dcluster_intervals = clustering.cluster_data_to_intervals(dcluster_data)\n",
    "dcluster_intervals.sort_values(by=[\"n_variants\"], ascending=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96ef086c",
   "metadata": {},
   "source": [
    "Apply filters to clustered variants (i.e., loci)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8fed4731",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove variants with less than 10 variants, and size between 800 and 8000 bp\n",
    "# this gives a theta range between 0.0125 and 0.00125\n",
    "dcluster_intervals_filtered = dcluster_intervals[\n",
    "    (dcluster_intervals[\"n_variants\"] >= 10) & \n",
    "    (dcluster_intervals[\"size\"] >= 800) & \n",
    "    (dcluster_intervals[\"size\"] <= 8000)\n",
    "].reset_index(drop=True)\n",
    "# add distance from preceeding loci\n",
    "dcluster_intervals_filtered = (\n",
    "    dcluster_intervals_filtered.groupby(\"chrom\").apply(\n",
    "        lambda df: clustering.distance_for_group(df),\n",
    "        include_groups=True,\n",
    "    ).reset_index(drop=True)\n",
    ")\n",
    "# to mitigate issues with linkage disequilibrium, remove loci with distance < 1000 bp\n",
    "dcluster_intervals_filtered = dcluster_intervals_filtered[ \n",
    "    (dcluster_intervals_filtered[\"distance\"] > 1000) | \n",
    "    (dcluster_intervals_filtered[\"distance\"] == 0)\n",
    "].reset_index(drop=True)\n",
    "dcluster_intervals_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a642cf85",
   "metadata": {},
   "outputs": [],
   "source": [
    "dcluster_intervals_filtered.to_csv(\n",
    "    \"../data/var/loci.bed\",\n",
    "    index=False,\n",
    "    header=False,\n",
    "    sep=\"\\t\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1291b6a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "admixed_samples = pd.read_csv(\"../data/var/admixture/admixed_individuals.csv\")\n",
    "csv_samples = \",\".join(admixed_samples[\"sample\"].tolist())\n",
    "csv_samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b8f8672",
   "metadata": {},
   "source": [
    "Produce fasta alignments for all loci."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "465a98fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "\n",
    "ref_path = \"../data/genome/reference.fasta.gz\"\n",
    "vcf_path = \"../data/var/filtered_variants.vcf.gz\"\n",
    "vcf_clean_path = vcf_path.replace(\".vcf.gz\", \"_clean.vcf.gz\")\n",
    "bed_path = \"../data/var/loci.bed\"\n",
    "\n",
    "subprocess.run([\n",
    "    \"bcftools\", \n",
    "    \"view\",\n",
    "    \"-s\", f\"^{csv_samples}\",\n",
    "    \"-Oz\",\n",
    "    \"-o\", vcf_clean_path,\n",
    "    vcf_path,\n",
    "    \"--write-index\",\n",
    "], check=True)\n",
    "\n",
    "subprocess.run([\n",
    "    \"vcf2fasta\", \n",
    "    \"--fasta\", ref_path,\n",
    "    \"--vcf\", vcf_clean_path,\n",
    "    \"--bed\", bed_path,\n",
    "    \"--out\", \"../data/phylo/loci\",\n",
    "    \"--force\",\n",
    "], check=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "conda",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
